### PR TITLE
secboot: read kernel efi image from snap file

### PR DESCRIPTION
**Pull request by cmatsuoka**
If a load sequence includes a snap file, assume it's the kernel snap
and create a SnapFileEFIImage load event that reads kernel.efi from
the squashfs instead of a regular FileEFIImage load event.
    
Signed-off-by: Claudio Matsuoka <claudio.matsuoka@canonical.com>

### CODE REVIEW

File: secboot/secboot_tpm_test.go [hunk 1]
```diff
@@ -494,6 +499,26 @@ func (s *secbootSuite) TestSealKey(c *C) {
...
 
+		// create a snap file
+		foo := filepath.Join(tmpDir, "foo.snap")
+		var foosnap snap.Container
+		if tc.badSnapFile {
+			err := ioutil.WriteFile(foo, nil, 0644)
+			c.Assert(err, IsNil)
+		} else {
+			snapdir := c.MkDir()
+			err := os.MkdirAll(filepath.Join(snapdir, "meta"), 0755)
```
Comment (mvo5) [inline secboot/secboot_tpm_test.go:510]: A little bit of duplication could be saved via something like:
```
snapYamlPath := filepath.Join(c.MkDir(), "meta/snap.yaml")
err := os.MkdirAll(filepath.Dir(snapYamlPath), 0755)
c.Assert(err, IsNil)
err = ioutil.WriteFile(filepath.Join(snapYamlPath, []byte("name: foo"), 0644)
c.Assert(err, IsNil)
...
```

File: secboot/secboot_tpm_test.go [hunk 2]
```diff
@@ -494,6 +499,26 @@ func (s *secbootSuite) TestSealKey(c *C) {
 			mockEFI[0] = "/does/not/exist"
 		}
 
+		// create a snap file
+		foo := filepath.Join(tmpDir, "foo.snap")
```
Comment (mvo5) [inline secboot/secboot_tpm_test.go:503]: Would it be more realistic to call this "kernel.snap" and build it with SnapType: "kernel" below? Probably makes not differences but AIUI we only use this for kernels, yes?

**Review by mvo5 — APPROVED** [review]
Looks good, thanks for this! Tiny suggestions/questions inline but no blockers.

File: secboot/secboot_tpm_test.go (hunk 1 above)
Comment (cmatsuoka) [reply secboot/secboot_tpm_test.go:510]: Did that and renamed variables to something a bit more descriptive.

**Review by zyga — APPROVED** [review]
Looks okay to me. Should we document that the file name must be `kernel.efi` anywhere? CC @degville

File: secboot/secboot_tpm.go [hunk 3]
```diff
@@ -445,14 +451,33 @@ func buildLoadSequences(pathSequences [][]string) []*sb.EFIImageLoadEvent {
 
 	loadEvents := make([]*sb.EFIImageLoadEvent, 0, len(pathSequences))
 
+	efiImage := func(name string) (sb.EFIImage, error) {
+		if strings.HasSuffix(name, ".snap") {
```
Comment (pedronis) [inline secboot/secboot_tpm.go:455]: my preference would be to be explicit about this and have something like:

```
type EFIImage interface  {  // to use in the params
   image()  sb.EFIImage
}

type FileEFIImage string

// ... implement image

type SnapEFIImage struct {
    Snap string
    Relative string
}

// ... implement image

```

Comment (cmatsuoka) [reply secboot/secboot_tpm.go:455]: Thanks. Will do something to make it not hardcoded in a follow-up.

**Pull request merged**